localstack/config.py

```diff
@@ -190,9 +191,7 @@ def load_environment(profile: str = None):
 
 def has_docker():
     try:
-        with open(os.devnull, "w") as devnull:
-            subprocess.check_output("docker ps", stderr=devnull, shell=True)
-        return True
+        return True if shutil.which("docker") else False
```

Comment:
Should we split that up into a `has_docker()` and `has_docker_cli()` function? Totally in favor of improving the startup speed 🚀 , but the semantics are a bit different (in some places we may want to determine whether the Docker engine is actually up and running). But for the purpose of determining `LAMBDA_EXECUTOR = "local"` below, the `has_docker_cli()` check should be sufficient.. 👍

Reply:
ok i understand the concern of this change and it should definitely not go into master like this.

tbh i'm wondering whether this check makes sense here at all. the separation of util functions into `has_docker_cli` and `is_docker_running` makes sense to me. it also makes sense to me that setting `LAMBDA_EXECUTOR=docker` should only be preferred if `is_docker_running` is true.

what doesn't make sense to me is making the determination of `is_docker_running` by calling `docker ps`, since we've removed the docker CLI from the container image. IMO we should be using the `DOCKER_CLIENT` for that. `has_docker_cli` is then pretty much only useful for checking whether we can use the `CmdDockerClient`, since the `SdkDockerClient` is preferred any way.

however we can't import DOCKER_CLIENT in `config.py`, so there's no obvious solution. i guess determining the LAMBDA_EXECUTOR mode should be part of the lambda service, or some lambda utils, and not of the config. unless we are using the LAMBDA_EXECUTOR flag elsewhere

Reply:
Great points. I think in general it makes sense to get the I/O calls out of the config. In this particular case we could only read `LAMBDA_EXECUTOR` from the environment, and then lazily evaluate the lambda executor to use as part of the Lambda executors logic. 👍